### PR TITLE
Toolchain+Ports+Kernel+LibC+LibCore+UE: Miscellaneous tweaks inspired by the mold port

### PR DESCRIPTION
**Pull request by BertalanD**
This PR fixes building the `mold` port with GCC, and adds two toolchain changes that will allow us to get rid of some patches in the future.

---

**Kernel+LibC+LibCore+UE: Implement fchmodat(2)**

This function is an extended version of `chmod(2)` that lets one control
whether to dereference symlinks, and specify a file descriptor to a
directory that will be used as the base for relative paths.

**Toolchain: Enable std::filesystem support in libc++**

Now that the last missing function (`fchmodat(2)`) has been added to
LibC, libc++ compiles successfully with std::filesystem support enabled.

**Toolchain: Enable the -rdynamic flag in the gcc driver**

Although we handle it in the linker spec file, gcc will actually reject
this argument unless it's also enabled in the option file.

CMake adds this flag if the minimum required version is 3.3 or less (see
CMP0065), so old projects would fail to compile because of this
unrecognized option

**LibC: Make the address argument of malloc_size a pointer to const**

We don't mutate the pointed-to memory, so let's be const correct.

Fixes building the `mimalloc` library that's optionally used by the mold
linker (note that it isn't enabled yet as I haven't tested it).

**Ports: Remove obsolete mold patch**

This is no longer needed, as both toolchains now support
std::filesystem.

### CODE REVIEW

File: Userland/Libraries/LibC/malloc.cpp [hunk 1]
```diff
@@ -500,7 +500,7 @@ void* calloc(size_t count, size_t size)
     return ptr;
 }
 
-size_t malloc_size(void* ptr)
+size_t malloc_size(const void* ptr)
```
Comment (ADKaster) [inline Userland/Libraries/LibC/malloc.cpp:503]: West const in 2022?

**Comment by ADKaster** [comment]
Very nice, only one nitpick. Always great to see one Posix extra boilerplate call enable a Toolchain feature.

**Review by linusg — COMMENTED** [review]
> but because the workaround is not too ugly let's not force people to rebuild their toolchain if they want to build mold.

Disagree. If we can remove it now, remove it now. It's not like everyone is building this port, and the people who are will need to accept the toolchain rebuild.

**Comment by BertalanD** [comment]
I removed the now-unnecessary patches and changed `const void*` to `void const*`.

**Pull request merged**